File: smoke_tests/dart/xss.dart
// XSS (Cross-Site Scripting) vulnerabilities in Dart

import 'dart:html';
import 'package:shelf/shelf.dart' as shelf;
import 'package:flutter_html/flutter_html.dart';

// Test 1: innerHTML assignment
void vulnerableInnerHtml(String userInput) {
  // VULNERABLE: Direct innerHTML assignment
  var element = document.getElementById('output');
  element?.innerHtml = userInput;
}

// Test 2: setInnerHtml without sanitizer
void vulnerableSetInnerHtml(String content) {
  // VULNERABLE: setInnerHtml without validator
  var element = DivElement();
  element.setInnerHtml(content);
}

// Test 3: document.write
void vulnerableDocumentWrite(String input) {
  // VULNERABLE: document.write with user input
  document.write(input);
}

// Test 4: Element creation with user content
void vulnerableCreateElement(String tagName, String content) {
  // VULNERABLE: User-controlled tag and content
  var element = Element.tag(tagName);
  element.text = content;
  document.body?.append(element);
}

// Test 5: Script element creation
void vulnerableScriptElement(String scriptContent) {
  // VULNERABLE: Creating script with user content
  var script = ScriptElement();
  script.text = scriptContent;
  document.head?.append(script);
}

// Test 6: Event handler attribute
void vulnerableEventHandler(String handler) {
  // VULNERABLE: User input in event handler
  var button = ButtonElement();
  button.setAttribute('onclick', handler);
}

// Test 7: href attribute injection
void vulnerableHrefInjection(String url) {
  // VULNERABLE: javascript: URL injection
  var link = AnchorElement();
  link.href = url;
  document.body?.append(link);
}

// Test 8: style attribute injection
void vulnerableStyleInjection(String style) {
  // VULNERABLE: CSS injection
  var element = DivElement();
  element.setAttribute('style', style);
}

// Test 9: Shelf response with HTML
shelf.Response vulnerableShelfResponse(String userContent) {
  // VULNERABLE: Unescaped HTML in response
  return shelf.Response.ok(
    '<html><body>$userContent</body></html>',
    headers: {'content-type': 'text/html'},
  );
}

// Test 10: Template literal injection
String vulnerableTemplate(String name) {
  // VULNERABLE: Template injection
  return '''
    <html>
      <body>
        <h1>Hello, $name!</h1>
      </body>
    </html>
  ''';
}

// Test 11: Flutter WebView HTML
void vulnerableWebViewHtml(String htmlContent) {
  // VULNERABLE: Loading user HTML in WebView
  // WebViewController().loadHtmlString(htmlContent);
  print('Would load: $htmlContent');
}

// Test 12: Flutter Html widget
Widget vulnerableHtmlWidget(String html) {
  // VULNERABLE: Rendering user HTML
  return Html(data: html);
}

// Test 13: JSON response with user data
shelf.Response vulnerableJsonResponse(String userData) {
  // VULNERABLE: User data in JSON that may be rendered
  return shelf.Response.ok(
    '{"message": "$userData"}',
    headers: {'content-type': 'application/json'},
  );
}

// Test 14: SVG injection
void vulnerableSvgInjection(String svgContent) {
  // VULNERABLE: SVG can contain scripts
  var container = DivElement();
  container.setInnerHtml(svgContent,
    treeSanitizer: NodeTreeSanitizer.trusted);
}

// Test 15: postMessage with user data
void vulnerablePostMessage(String message) {
  // VULNERABLE: Cross-origin message with user data
  window.postMessage(message, '*');
}

File: smoke_tests/swift/xss.swift
// Cross-Site Scripting (XSS) vulnerabilities in Swift (WebView contexts)
import Foundation
import WebKit

class XssVulnerabilities {
    var webView: WKWebView!

    // Test 1: loadHTMLString with user input
    func loadUserContent(userInput: String) {
        // VULNERABLE: User input directly in HTML
        let html = "<html><body><h1>\(userInput)</h1></body></html>"
        webView.loadHTMLString(html, baseURL: nil)
    }

    // Test 2: JavaScript execution with user data
    func executeUserScript(userData: String) {
        // VULNERABLE: User data in JavaScript
        let script = "document.getElementById('output').innerHTML = '\(userData)';"
        webView.evaluateJavaScript(script, completionHandler: nil)
    }

    // Test 3: URL loading without validation
    func loadUrl(urlString: String) {
        // VULNERABLE: Can load javascript: URLs
        if let url = URL(string: urlString) {
            webView.load(URLRequest(url: url))
        }
    }

    // Test 4: JavaScript enabled for untrusted content
    func configureWebView() {
        let config = WKWebViewConfiguration()
        let prefs = WKPreferences()
        // VULNERABLE: JavaScript enabled for all content
        prefs.javaScriptEnabled = true
        config.preferences = prefs
        webView = WKWebView(frame: .zero, configuration: config)
    }

    // Test 5: Message handler without sanitization
    func setupMessageHandler() {
        let contentController = WKUserContentController()
        // Handler receives user-controlled data from JavaScript
        contentController.add(self, name: "messageHandler")
    }

    // Test 6: User script injection
    func injectUserScript(code: String) {
        // VULNERABLE: Injecting user-controlled script
        let script = WKUserScript(source: code, injectionTime: .atDocumentEnd, forMainFrameOnly: false)
        webView.configuration.userContentController.addUserScript(script)
    }

    // Test 7: String interpolation in JavaScript
    func updateElement(elementId: String, content: String) {
        // VULNERABLE: Both parameters from user
        let js = "document.getElementById('\(elementId)').innerHTML = '\(content)';"
        webView.evaluateJavaScript(js, completionHandler: nil)
    }

    // Test 8: JSON embedding without encoding
    func embedJsonData(data: [String: Any]) {
        if let jsonData = try? JSONSerialization.data(withJSONObject: data),
           let jsonString = String(data: jsonData, encoding: .utf8) {
            // VULNERABLE: JSON data may contain XSS payload
            let html = "<html><body><script>var data = \(jsonString);</script></body></html>"
            webView.loadHTMLString(html, baseURL: nil)
        }
    }

    // Test 9: Template rendering
    func renderTemplate(name: String, value: String) {
        // VULNERABLE: User values in template
        let html = """
        <html>
        <body>
            <h1>Hello, \(name)!</h1>
            <p>Your value: \(value)</p>
        </body>
        </html>
        """
        webView.loadHTMLString(html, baseURL: nil)
    }

    // Test 10: URL parameter reflection
    func loadWithParams(baseUrl: String, param: String) {
        // VULNERABLE: Parameter reflected in page
        let html = "<html><body><script>var param = '\(param)';</script></body></html>"
        webView.loadHTMLString(html, baseURL: URL(string: baseUrl))
    }

    // Test 11: Event handler injection
    func addClickHandler(handler: String) {
        // VULNERABLE: User-controlled event handler
        let html = "<html><body><button onclick=\"\(handler)\">Click</button></body></html>"
        webView.loadHTMLString(html, baseURL: nil)
    }

    // Test 12: Style injection
    func applyUserStyle(style: String) {
        // VULNERABLE: CSS injection can lead to data exfiltration
        let html = "<html><head><style>\(style)</style></head><body>Content</body></html>"
        webView.loadHTMLString(html, baseURL: nil)
    }
}

// WKScriptMessageHandler conformance
extension XssVulnerabilities: WKScriptMessageHandler {
    func userContentController(_ userContentController: WKUserContentController,
                              didReceive message: WKScriptMessage) {
        // Handle message
    }
}
